helix-core/src/syntax/tree_cursor.rs

```diff
@@ -0,0 +1,114 @@
+use std::{collections::HashMap, ops::Range};
+
+use super::{LanguageLayer, LayerId};
+
+use slotmap::HopSlotMap;
+use tree_sitter::Node;
+
+pub struct TreeCursor<'a> {
+    layers: &'a HopSlotMap<LayerId, LanguageLayer>,
+    root: LayerId,
+    current: LayerId,
+    injection_ranges: HashMap<Range<usize>, LayerId>,
+    // TODO: Ideally this would be a `tree_sitter::TreeCursor<'a>` but
+    // that returns very surprising results in testing.
+    cursor: Node<'a>,
```

Comment:
I planned to use `tree_sitter::TreeCursor` here but it gave obviously wrong results, for example `TreeCursor::goto_parent` returning `false` (indicating there is no parent) when `cursor.node().parent().is_some()` and `TreeCursor::goto_next_sibling` returning false when there were next siblings. I might not be understanding the API fully or it may have bugs. I don't see it as a blocker though since we currently use `Node::parent`/`Node::next_sibling`/`Node::prev_sibling` in `A-poin`.